code/crates/sync/src/handle.rs

```diff
@@ -619,80 +703,251 @@ where
+    }
+
+    #[test]
+    fn stops_before_next_blocking_range() {
+        let out = compute_free_window(H(9), H(100), 10, vec![range(15, 20)]).unwrap();
+        assert_eq!(*out.start(), H(10));
+        assert_eq!(*out.end(), H(14));
+    }
+
+    #[test]
+    fn jumps_over_overlapping_blocks() {
+        let out = compute_free_window(
+            H(9),
+            H(100),
+            3,
+            vec![range(10, 12), range(11, 13), range(13, 14)],
+        )
+        .unwrap();
+        assert_eq!(*out.start(), H(15));
+        assert_eq!(*out.end(), H(17));
+    }
+
+    #[test]
+    fn peer_cap_hits_inside_growth() {
+        let out = compute_free_window(H(0), H(5), 100, vec![]).unwrap();
+        assert_eq!(*out.start(), H(1));
+        assert_eq!(*out.end(), H(5));
+    }
+
+    #[test]
+    fn hole_between_blocks_fills_up_to_batch_limit() {
+        let out = compute_free_window(H(0), H(100), 4, vec![range(1, 3), range(10, 99)]).unwrap();
+        assert_eq!(*out.start(), H(4));
+        assert_eq!(*out.end(), H(7));
+    }
+
+    #[test]
+    fn hole_smaller_than_batch_stops_at_block() {
+        let out = compute_free_window(H(0), H(100), 10, vec![range(1, 3), range(9, 20)]).unwrap();
+        assert_eq!(*out.start(), H(4));
+        assert_eq!(*out.end(), H(8));
+    }
+
+    // helper to normalize `Vec<RangeInclusive<H>>` for easy equality checks
+    fn as_pairs(v: &[RangeInclusive<H>]) -> Vec<(u64, u64)> {
+        let mut pairs: Vec<_> = v
+            .iter()
+            .map(|r| (r.start().as_u64(), r.end().as_u64()))
+            .collect();
+        pairs.sort_unstable();
+        pairs
+    }
+
+    #[test]
+    fn not_found_returns_err() {
+        let mut v = vec![range(5, 7), range(9, 12)];
+        let err = excise_height(&mut v, H(8)).unwrap_err();
+        assert!(err.contains("cannot find range"), "got: {}", err);
+        assert_eq!(as_pairs(&v), vec![(5, 7), (9, 12)]);
+    }
+
+    #[test]
+    fn singleton_range_removed() {
+        let mut v = vec![range(5, 5)];
+        excise_height(&mut v, H(5)).unwrap();
+        assert!(v.is_empty());
+    }
+
+    #[test]
+    fn singleton_range_removed_with_other_ranges() {
+        let mut v = vec![range(1, 3), range(5, 5), range(9, 11)];
+        excise_height(&mut v, H(5)).unwrap();
+        assert_eq!(as_pairs(&v), vec![(1, 3), (9, 11)]);
+    }
+
+    #[test]
+    fn height_at_left_edge_trims_left() {
+        let mut v = vec![range(5, 10)];
+        excise_height(&mut v, H(5)).unwrap();
+        assert_eq!(as_pairs(&v), vec![(6, 10)]);
+    }
+
+    #[test]
+    fn height_at_right_edge_trims_right() {
+        let mut v = vec![range(5, 10)];
+        excise_height(&mut v, H(10)).unwrap();
+        assert_eq!(as_pairs(&v), vec![(5, 9)]);
+    }
+
+    #[test]
+    fn height_inside_splits_into_two() {
+        let mut v = vec![range(5, 10)];
+        excise_height(&mut v, H(7)).unwrap();
+        assert_eq!(as_pairs(&v), vec![(5, 6), (8, 10)]);
+    }
```

Comment:
it seems like you're trying to run different testcases for the same scenario. maybe you want to use [rstest](https://docs.rs/rstest/0.26.1/rstest/attr.rstest.html#test-parametrized-cases) crate.